### PR TITLE
⭐️ nmap provider

### PR DESCRIPTION
**Pull request by chris-rock**
# nmap provider

Nmap, short for Network Mapper, is a powerful and versatile open-source tool used for network discovery and security auditing. This tool is widely utilized by network administrators, security professionals, and penetration testers to map out network structures, discover hosts, identify services, and detect vulnerabilities.

The nmap provider maps primary objects and attributes that nmap uses to store and manage information about scanned targets, discovered hosts, and their associated ports and services.

## Pre-requisites

This provider requires the nmap tool to be installed on your system. You can download and install nmap from the official [website](https://nmap.org/download.html).

## Get Started

```shell
cnquery shell nmap
```

## Example

*Scan active IP address in network*

```shell
nmap.network("192.168.178.0/24").hosts { name ports { * }  }
nmap.network.hosts: [
  0: {
    ports: [
      0: {
        service: "http"
        version: ""
        method: "probed"
        state: "open"
        protocol: "tcp"
        port: 443
        product: "FRITZ!Box http config"
      }
      1: {
        service: "sip"
        version: ""
        method: "probed"
        state: "open"
        protocol: "tcp"
        port: 5060
        product: "AVM FRITZ!OS SIP"
      }
    ]
    name: "192.168.178.1"
  }
  1: {
    ports: [
      0: {
        service: "rtsp"
        version: "770.8.1"
        method: "probed"
        state: "open"
        protocol: "tcp"
        port: 5000
        product: "AirTunes rtspd"
      }
      1: {
        service: "rtsp"
        version: "770.8.1"
        method: "probed"
        state: "open"
        protocol: "tcp"
        port: 7000
        product: "AirTunes rtspd"
      }
    ]
    name: "192.168.178.25"
  }
]
```

*Host scan with specific ip*

```shell
nmap.network(target: "192.168.178.25").hosts { ports }
nmap.network.hosts: [
  0: {
    ports: [
      0: nmap.port port=5000 service="rtsp"
      1: nmap.port port=7000 service="rtsp"
    ]
  }
]
```

# Advanced Usage

Discover all exposed hosts on a network.

```shell
cnquery shell nmap --networks "192.168.0.0/20" --discover hosts
```

Connect to a specific IP address and display all open ports.

```shell
cnquery shell nmap host 8.8.8.8
```

## Verifying the Installation of nmap

To verify the installation of nmap, run the following command:

```shell
cnquery run nmap -c "nmap.version { * }"
nmap.version: {
  compiledWithout: []
  nsockEngines: [
    0: "kqueue"
    1: "poll"
    2: "select"
  ]
  version: "7.95"
  platform: "arm-apple-darwin23.4.0"
  compiledWith: [
    0: "liblua-5.4.6"
    1: "openssl-3.3.1"
    2: "libssh2-1.11.0"
    3: "libz-1.2.12"
    4: "libpcre2-10.44"
    5: "nmap-libpcap-1.10.4"
    6: "nmap-libdnet-1.12"
    7: "ipv6"
  ]
}
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -508,6 +518,11 @@ lr/docs/markdown: providers/lr
 		--description "The Network resource pack lets you use MQL to query and assess the security of domains and network services." \
 		--docs-file providers/network/resources/network.lr.manifest.yaml \
 		--output ../docs/docs/mql/resources/network-pack
+	./lr markdown providers/network/resources/nmap.lr \
+		--pack-name "nmap" \
+		--description "The nmap resource pack lets you use MQL to query and assess nmap data." \
```
Comment (tas50) [inline Makefile]:
```suggestion
		--description "The Nmap resource pack lets you use MQL to query and assess Nmap data." \
```

File: providers/defaults.go [hunk 2]
```diff
@@ -230,6 +230,21 @@ var DefaultProviders Providers = map[string]*Provider{
...
+	"nmap": {
+		Provider: &plugin.Provider{
+			Name:            "nmap",
+			ID:              "go.mondoo.com/cnquery/v9/providers/nmap",
+			ConnectionTypes: []string{"nmap"},
+			Connectors: []plugin.Connector{
+				{
+					Name:  "nmap",
+					Use:   "nmap",
+					Short: "a nmap network scanner",
```
Comment (tas50) [inline providers/defaults.go]:
```suggestion
					Short: "a Nmap network scanner",
```

File: providers/nmap/README.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
+# nmap provider
```
Comment (tas50) [inline providers/nmap/README.md]:
```suggestion
# Nmap provider
```

File: providers/nmap/README.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+Nmap, short for Network Mapper, is a powerful and versatile open-source tool used for network discovery and security auditing. This tool is widely utilized by network administrators, security professionals, and penetration testers to map out network structures, discover hosts, identify services, and detect vulnerabilities.
+
+The nmap provider maps primary objects and attributes that nmap uses to store and manage information about scanned targets, discovered hosts, and their associated ports and services.
+
+## Pre-requisites
+
+This provider requires the nmap tool to be installed on your system. You can download and install nmap from the official [website](https://nmap.org/download.html).
```
Comment (tas50) [inline providers/nmap/README.md]:
```suggestion
This provider requires the Nmap tool to be installed on your system. You can download and install Nmap from the official [website](https://nmap.org/download.html).
```

File: providers/nmap/config/config.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+var Config = plugin.Provider{
+	Name:            "nmap",
+	ID:              "go.mondoo.com/cnquery/v11/providers/nmap",
+	Version:         "11.0.0",
+	ConnectionTypes: []string{provider.DefaultConnectionType},
+	Connectors: []plugin.Connector{
+		{
+			Name:      "nmap",
+			Use:       "nmap",
+			Short:     "a nmap network scanner",
```
Comment (tas50) [inline providers/nmap/config/config.go]:
```suggestion
			Short:     "a Nmap network scanner",
```

**Comment by chris-rock** [comment]
For discovery:

```
nmap -sT -T4 192.184.215.14
```

For service detection:

```
nmap -sT -T4 -sV 93.184.215.14
```

**Comment by atomic111** [comment]
just to do a very fast ip discovery without scanning any port:

```
nmap -sn -n --min-parallelism 100 -T4 192.168.1.0/24
```

**Comment by atomic111** [comment]
make a port scan just for the discovered hosts/ips

```
nmap -sT -sV  -n --min-parallelism 100 -T4 192.168.1.0/24
```

**Review by atomic111 — APPROVED** [review]
@chris-rock awesome. thank you for the new nmap provider

**Pull request merged**